password_security/models/res_users.py

```diff
@@ -52,41 +80,17 @@ def _check_password_rules(self, password):
         if not password:
             return True
         company_id = self.company_id
-        message = []
-        if company_id.password_lower and sum(map(str.islower, password)) < \
-                company_id.password_lower:
-            message.append('\n ' + _('Lowercase letter (At least ' +
-                                     str(company_id.password_lower) +
-                                     ' character)')
-                           )
-        if company_id.password_upper and sum(map(str.isupper, password)) < \
-                company_id.password_upper:
-            message.append('\n ' + _('Uppercase letter (At least ' +
-                                     str(company_id.password_upper) +
-                                     ' character)')
-                           )
-        if company_id.password_numeric and sum(map(str.isdigit, password)) < \
-                company_id.password_numeric:
-            message.append('\n ' + _('Numeric digit (At least ' +
-                                     str(company_id.password_numeric) +
-                                     ' numeric)')
-                           )
-        if company_id.password_special and len(set('[~!@#$%^&*()_+{}":;\']+$'
-                                                   ).intersection(
-                password)) < company_id.password_numeric:
-            message.append('\n ' + _('Special character (At least ' +
-                                     str(company_id.password_special) +
-                                     ' character of [ ~ ! @ # $ % ^ & * ( )_+ '
-                                     '{ } " : ; \' ])')
-                           )
-        if company_id.password_length and len(password) < \
-                company_id.password_length:
-            message = [_('Password must be %d characters or more.') %
-                       company_id.password_length] + message
-        if len(message) > 0:
-            raise PassError('\r'.join(message))
-        else:
-            return True
+        password_regex = [
+            '^',
+            '(?=.*?[a-z]){' + str(company_id.password_lower) + ',}',
```

Comment:
Actually you can strip the `int()`:

```python
>>> "%d" % False
'0'
```

Reply:
Oh cool TIL- I didn't know that the `%d` would typecast non-ints. I thought it threw an error like it will with a string:

```
>>> '%d' % '1'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: %d format: a number is required, not str

```

Reply:
Actually that's exactly what it does 😆

Reply:
Yeah the `%d` is meant to serve as a validation as well, so it fails with things that don't have a direct int val. I guess `False` does technically have a direct int val though, because 0 or 1

Reply:
`bool` actually inherits from `int`:

```
>>> issubclass(bool, int)
True
```

Reply:
And that's a TIL for me! Today is a good day for learning 😄